File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        leftP = 0
        rightP = len(matrix) - 1

        while leftP <= rightP:
            middle = (rightP + leftP) // 2
            currList = matrix[middle]

            if target >= currList[0] and target <= currList[len(currList)-1]:
                lp = 0
                rp = len(currList) - 1
                while lp <= rp:
                    mid = (lp + rp) // 2
                    if currList[mid] == target:
                        return True
                    if currList[mid] < target:
                        lp = mid + 1
                    if currList[mid] > target:
                        rp = mid = 1
                return False 

            if target < currList[0]:
                rightP = middle - 1
            if target > currList[len(currList)-1]:
                leftP = middle + 1

        return False 

File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int outterLeft = 0;
        int outterRight = matrix.length - 1;

        while (outterLeft <= outterRight) {
            int outterMiddle = outterLeft + ((outterRight - outterLeft) / 2);
            int[] currMatrix = matrix[outterMiddle];

            if (target >= currMatrix[0] && target <= currMatrix[currMatrix.length - 1]){
                int innerLeft = 0;
                int innerRight = currMatrix.length - 1;

                while (innerLeft <= innerRight){
                    int innerMiddle = innerLeft + ((innerRight - innerLeft) / 2);
                    if (currMatrix[innerMiddle] == target){
                        return true;
                    } else if (target > currMatrix[innerMiddle]  ){
                        innerLeft = innerMiddle + 1;
                    } else {
                        innerRight = innerMiddle - 1;
                    }
                }
                return false;

            } else if (target < currMatrix[0]){
                outterRight = outterMiddle - 1;
            } else{
                outterLeft = outterMiddle + 1;
            }
        }

        return false;

    }
}
